crates/wasmtime/src/runtime/vm/continuation.rs

```diff
@@ -293,84 +293,6 @@ pub mod optimized {
-                    "Attempt to dereference null VMContRef!"
-                ))
-            })?
-        };
-        assert!(cont.state == State::Allocated || cont.state == State::Invoked);
-
-        if ENABLE_DEBUG_PRINTING {
-            let chain = instance.typed_continuations_stack_chain();
-            // SAFETY: We maintain as an invariant that the stack chain field in the
-            // VMContext is non-null and contains a chain of zero or more
-            // StackChain::Continuation values followed by StackChain::Main.
-            match unsafe { (**chain).0.get_mut() } {
-                StackChain::Continuation(running_contref) => {
-                    debug_assert_eq!(contref, *running_contref);
-                    debug_println!(
-                        "Resuming contref @ {:p}, previously running contref is {:p}",
-                        contref,
-                        running_contref
-                    )
-                }
-                _ => {
-                    // Before calling this function as a libcall, we must have set
-                    // the parent of the to-be-resumed continuation to the
-                    // previously running one. Hence, we must see a
-                    // `StackChain::Continuation` variant.
-                    return Err(TrapReason::user_without_backtrace(anyhow::anyhow!(
-                        "Invalid StackChain value in VMContext"
-                    )));
-                }
-            }
-        }
-
-        Ok(cont.stack.resume())
-    }
-
-    /// TODO
-    #[inline(always)]
-    pub fn suspend(instance: &mut Instance, tag_addr: *mut u8) -> Result<(), TrapReason> {
-        let chain_ptr = instance.typed_continuations_stack_chain();
-
-        // TODO(dhil): This should be handled in generated code.
-        // SAFETY: We maintain as an invariant that the stack chain field in the
-        // VMContext is non-null and contains a chain of zero or more
-        // StackChain::Continuation values followed by StackChain::Main.
-        let chain = unsafe { (**chain_ptr).0.get_mut() };
-        let running = match chain {
-            StackChain::Absent => Err(TrapReason::user_without_backtrace(anyhow::anyhow!(
-                "Internal error: StackChain not initialised"
-            ))),
-            StackChain::MainStack { .. } => Err(TrapReason::user_without_backtrace(
-                anyhow::anyhow!("Calling suspend outside of a continuation"),
-            )),
-            StackChain::Continuation(running) => {
-                // SAFETY: See above.
-                Ok(unsafe { &**running })
-            }
-        }?;
-
-        let stack = &running.stack;
-        debug_println!(
-            "Suspending while running {:p}, parent is {:?}",
-            running,
-            running.parent_chain
-        );
-
-        let payload = ControlEffect::suspend(tag_addr as *const u8);
-        Ok(stack.suspend(payload))
-    }
-
```

Comment:
I wonder whether there may be value in keeping this implementation around, perhaps as a new baseline implementation...

Reply:
Are you suggesting to replace the current baseline implementation with the "optimized" implementation prior to this PR? I definitely wouldn't want to maintain a third active implementation.

Reply:
That would require keeping the version of the `fibre/*` files prior to this PR around (effectively continuing to support a third possible internal layout of stacks), plus eventually increasing divergence within `optimized.rs`. I consider the maintenance cost of that too great.
If we want to evaluate the performance of the implementation prior to this PR, we should use the approach I've implemented where we checkout older revisions directly, accepting that it will not support things like `switch` in the future.

Reply:
I don't think it is such a burden. We should discuss this on Thursday, I think.

Reply:
Okay, let's discuss on Thursday. I will merge this PR in the meantime, since making the previous optimized implementation available as a third implementation would be a PR of its own.

Reply:
Yes, agreed.